File: src/main/kotlin/com/xs0/gqlktx/types/gql/GInterfaceType.java
package com.xs0.gqlktx.types.gql;

import com.xs0.gqlktx.QueryException;
import com.xs0.gqlktx.schema.builder.TypeKind;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;

import java.util.Map;
import java.util.Set;

public class GInterfaceType extends GFieldedType {
    private Set<GObjectType> implementations;

    public GInterfaceType(String name, Map<String, GField> fields) {
        super(name, fields);
    }

    @Override
    public TypeKind getKind() {
        return TypeKind.INTERFACE;
    }

    @Override
    public boolean validAsArgumentType() {
        return false;
    }

    public void setImplementations(Set<GObjectType> implementations) {
        this.implementations = implementations;

        for (GObjectType objectType : implementations) {
            objectType.addInterface(this);
        }
    }

    public Set<GObjectType> getImplementations() {
        return implementations;
    }

    @Override
    public void coerceValue(JsonObject raw, String key, JsonObject out) throws QueryException {
        throw new QueryException("Interface type " + getName() + " can't be used as a variable");
    }

    @Override
    public void coerceValue(JsonArray raw, int index, JsonArray out) throws QueryException {
        throw new QueryException("Interface type " + getName() + " can't be used as a variable");
    }

    @Override
    public void toString(StringBuilder sb) {
        sb.append("interface ").append(getName()).append(" {\n");
        dumpFieldsToString(sb);
        sb.append("}\n");
    }
}

File: src/main/kotlin/com/xs0/gqlktx/types/gql/GInterfaceType.kt
package com.xs0.gqlktx.types.gql

import com.xs0.gqlktx.QueryException
import com.xs0.gqlktx.dom.Value
import com.xs0.gqlktx.schema.builder.TypeKind
import com.xs0.gqlktx.schema.intro.GqlIntroType
import java.util.ArrayList
import java.util.LinkedHashSet

class GInterfaceType(name: String, fields: Map<String, GField>, description: String?) : GFieldedType(name, fields, description) {
    lateinit var implementations: Set<GFieldedType>
        private set

    fun setImpls(implementations: Set<GFieldedType>) {
        this.implementations = implementations

        for (objectType in implementations) {
            objectType.addInterface(this)
        }
    }

    val implsForIntrospection: List<GqlIntroType> by lazy {
        val res = ArrayList<GqlIntroType>(implementations.size)
        for (posib in implementations)
            res.add(posib.introspector)
        res
    }

    override val kind: TypeKind
        get() = TypeKind.INTERFACE

    override fun coerceValue(raw: Value): Value {
        throw QueryException("Interface type $name can't be used as a variable")
    }

    override fun toString(sb: StringBuilder) {
        sb.append("interface ").append(name).append(" {\n")
        dumpFieldsToString(sb)
        sb.append("}\n")
    }
}
